src/Utilities/SingleStackUtils/SingleStackUtils.jl

```diff
@@ -453,4 +454,258 @@ function dict_of_nodal_states(
+    i::TI
+    j::TJ
+    cart_ind::CI
+    interp::IN
+    function GridStack(
+        bl::BalanceLaw,
+        grid::DiscontinuousSpectralElementGrid,
+        prognostic,
+        auxiliary,
+        diffusive,
+        hyperdiffusive;
+        i = 1,
+        j = 1,
+        interp = true,
+    )
+        states = States(prognostic, auxiliary, diffusive, hyperdiffusive)
+        vrange = 1:size(prognostic, 3)
+        N = polynomialorders(grid)
+        dim = dimensionality(grid)
+        Nq = N .+ 1
+        Nqk = dim == 2 ? 1 : Nq[dim]
+        Nk = dim == 2 ? 1 : N[dim]
+        # Store cartesian indices, so we can map the iter_state
+        # to the cartesian space `Q[i, var, j]`
+        if interp
+            cart_ind = CartesianIndices((Nk, size(prognostic, 3)))
+        else
+            cart_ind = CartesianIndices((Nqk, size(prognostic, 3)))
+        end
+        args = (bl, grid, states, vrange, i, j, cart_ind, interp)
+        BL, G, S, VR, TI, TJ, CI, IN = typeof.(args)
+        if interp
+            len = size(prognostic, 3) * Nk + 1
+        else
+            len = size(prognostic, 3) * Nqk
+        end
+        new{len, BL, G, S, VR, TI, TJ, CI, IN}(args...)
+    end
+end
+
+# Convenience wrapper
+GridStack(solver_config; kwargs...) = GridStack(
+    solver_config.dg.balance_law,
+    solver_config.dg.grid,
+    solver_config.Q,
+    solver_config.dg.state_auxiliary,
+    solver_config.dg.state_gradient_flux,
+    solver_config.dg.states_higher_order[2];
+    kwargs...,
+)
+
+Base.length(gs::GridStack{N}) where {N} = N
+
+# Helper function
+get_state(v, state, vid⁻, vid⁺, ev⁻, ev⁺, J⁻, J⁺) =
+    (J⁻ * state[vid⁻, v, ev⁻] + J⁺ * state[vid⁺, v, ev⁺]) / (J⁻ + J⁺)
+
+to_cpu(state) =
+    array_device(state) isa CPU ? state.realdata : Array(state.realdata)
+
+function interp_top(state, args, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> get_state(v, state, args...), 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function interp_bot(state, vid⁻, ev⁻, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> state[vid⁻, v, ev⁻], 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function no_interp(state, ijk, ev, n_vars, bl, st)
+    vs = Vars{vars_state(bl, st, eltype(state))}
+    if n_vars ≠ 0
+        return vs(map(v -> state[ijk, v, ev], 1:n_vars))
+    else
+        return nothing
+    end
+end
+
+function Base.iterate(gs::GridStack, iter_state = 1)
+    iter_state > length(gs) && return nothing
+
+    # extract grid information
+    grid = gs.grid
+    FT = eltype(grid)
+    N = polynomialorders(grid)
+    dim = dimensionality(grid)
+    Nq = N .+ 1
+    # Code assumes the same polynomial order in all horizontal directions
+    @inbounds begin
+        Nq1 = Nq[1]
+        Nq2 = Nq[2]
+        Nqk = dim == 2 ? 1 : Nq[dim]
+    end
+    Np = dofs_per_element(grid)
```

Comment:
Ditto.

Also: is there a plan/expectation/need to support different polynomial orders in horizontal directions? If there is, then we shouldn't be adding code that doesn't support it, and if there isn't, then there's no need for the comment and why  do we have a way to do such a thing?